<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>API for HOMER in R for Genomic Analysis using Tidy Conventions • marge</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="API for HOMER in R for Genomic Analysis using Tidy Conventions">
<meta property="og:description" content="API for HOMER in R for genomic analysis using tidy conventions. Runs HOMER from native R functions and allows for organization of motif enrichment analysis, and subsequent reading in of motifs into tidily formatted tibbles for further analysis and visualization of many parallel analyses.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">marge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4.9999</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/marge-workflow.html">`marge`: An API for Analysis of Motifs Using HOMER in R</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="marge-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#marge-" class="anchor"></a>marge <img src="reference/figures/Marge_Simpson.png" align="right" height="205" width="132">
</h1></div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The aim of <code>marge</code> is to provide an R API to <code>HOMER</code> for the analysis of genomic data, utilizing a tidy framework to accelerate organization and visualization of analyses.</p>
</div>
<div id="installing-homer" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-homer" class="anchor"></a>Installing <code>HOMER</code>
</h2>
<p>First, running <code>marge</code> requires having a working installation of <code>HOMER</code> on your computer. Please see the <a href="http://homer.ucsd.edu/homer/">HOMER website</a> for more information on installing and configuring <code>HOMER</code> and to learn more about the methodology. In particular, note that you should install your desired genomes in addition to installing <code>HOMER</code> using the <code>./configureHomer.pl</code> script.</p>
<p>Note that working with a conda installation of <code>HOMER</code> is not well tested at this time. A potential workaround is below. I recommend installing directly from source.</p>
</div>
<div id="installing-marge" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-marge" class="anchor"></a>Installing <code>marge</code>
</h2>
<p>To install the latest development version of <code>marge</code>, simply do:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">install_github</a></span>(<span class="st">'robertamezquita/marge'</span>, <span class="dt">ref =</span> <span class="st">'master'</span>)</span></code></pre></div>
</div>
<div id="before-running-marge" class="section level2">
<h2 class="hasAnchor">
<a href="#before-running-marge" class="anchor"></a>Before Running <code>marge</code>
</h2>
<p>While <code>marge</code> will do its best to find <code>HOMER</code>, there are certain environments where it will not be able to do so, specifically, with regards to RStudio and conda installs of <code>HOMER</code>. In these cases, a custom path to <code>HOMER</code> can be provided if it is not found by the package’s utilities by setting <code><a href="https://rdrr.io/r/base/options.html">options('homer_path' = "/path/to/homer-4.10")</a></code>. This can be set in your <code>~/.Rprofile</code> so it loads the correct path automagically each time.</p>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p><code>marge</code> is currently semi-stable. The package currently includes the ability to:</p>
<ul>
<li>Run a motif analysis: <code><a href="reference/find_motifs_genome.html">find_motifs_genome()</a></code> - runs the <code>HOMER</code> script <code>findMotifsGenome.pl</code> via R, and outputs a results directory in the default <code>HOMER</code> style</li>
<li>Read in results: <code>read_*_results()</code> - read in either <code>denovo</code> or <code>known</code> enriched motifs with the <code><a href="reference/read_denovo_results.html">read_denovo_results()</a></code> or <code><a href="reference/read_known_results.html">read_known_results()</a></code> functions, pointing to the <code>HOMER</code> directory that was created in the previous step. The <code>read_*</code> functions produce tibbles summarizing the motif enrichment results into a tidy format for easier visualization and analysis. See the reference pages of each for more details.</li>
<li>Write motifs in HOMER compatible format with <code><a href="reference/write_homer_motif.html">write_homer_motif()</a></code>
</li>
<li>Find specific motif instances across regions using supplied PWMs with <code><a href="reference/find_motifs_instances.html">find_motifs_instances()</a></code> and read in the results with <code><a href="reference/read_motifs_instances.html">read_motifs_instances()</a></code>
</li>
<li>Access the HOMER database of known motifs by inspecting the <code>HOMER_motifs</code> object</li>
</ul>
<p>Further details can be found in the associated vignette, describing installation and typical workflows encompassing basic/advanced usage schemas.</p>
</div>
<div id="compared-to-homer-alone" class="section level2">
<h2 class="hasAnchor">
<a href="#compared-to-homer-alone" class="anchor"></a>Compared to HOMER alone</h2>
<p>Like the actual Homer Simpson, <code>HOMER</code> is made better with the addition of <code>marge</code>. With the continually increasing throughput in conducting sequencing analysis, <code>marge</code> provides a native R framework to work from end to end with motif analyses - from processing to storing to visualizing these results, all using modern tidy conventions.</p>
<hr>
<div id="contact" class="section level4">
<h4 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h4>
<ul>
<li>Robert A. Amezquita (aut, cre, cph): <a href="robert.amezquita@fredhutch.org">robert.amezquita@fredhutch.org</a>
</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Robert A. Amezquita <br><small class="roles"> Author, maintainer, copyright holder </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Robert A. Amezquita.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
