<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
<code>marge</code>: An API for Analysis of Motifs Using HOMER in R • marge</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="&lt;code&gt;marge&lt;/code&gt;: An API for Analysis of Motifs Using HOMER in R">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">marge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/marge-workflow.html">`marge`: An API for Analysis of Motifs Using HOMER in R</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>
<code>marge</code>: An API for Analysis of Motifs Using HOMER in R</h1>
                        <h4 class="author">Robert A. Amezquita</h4>
            
            <h4 class="date">2019-01-03</h4>
      
      
      <div class="hidden name"><code>marge-workflow.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Profiling of open chromatin regions through the assay for transposase-accessible chromatin (ATAC) and transcription factor binding via chromatin immunoprecipitation (ChIP) sequencing has increased our ability to resolve patterns of putative transcription factor binding sites at the genome-wide level. Popular tools such as <a href="http://homer.ucsd.edu/homer/">HOMER</a> and <a href="http://meme-suite.org/">MEME</a> have driven forward the analyses of sequence composition, deriving <em>de novo</em> motifs and searching for the enrichment of known motifs. However, their interfaces do not allow for the construction of parallel inquiries of multiple datasets. Furthermore, their results do not conform to formats amenable to ‘tidy’ analyses, presenting a significant bottleneck in motif analysis. Here, I present <code>marge</code>, a companion <code>R</code> package that interfaces with HOMER to facilitate the construction of queries and to tidy results for further downstream analyses.</p>
</div>
<div id="methods" class="section level1">
<h1 class="hasAnchor">
<a href="#methods" class="anchor"></a>Methods</h1>
<div id="implementation" class="section level2">
<h2 class="hasAnchor">
<a href="#implementation" class="anchor"></a>Implementation</h2>
<p><code>marge</code> is an <code>R</code> package that makes use of various <code>tidyverse</code> conventions, allowing for modern conventions of <code>R</code> programming and manipulations. <code>marge</code> interfaces directly with HOMER to construct queries, execute motif enrichment searches, and organize results. While this requires a separate installation of HOMER and introduces this long-term dependency, core HOMER routines have been stable for a long time, with infrequent updates. Given the popularity of HOMER, the syntax and construction of queries will thus be familiar to the audience utilizing the <code>marge</code> companion package. However, to conform to modern standards of software design, certain core routines, such as <code>findMotifsGenome.pl</code> which encompasses many different tasks, have been partitioned into distinct <code>R</code> functions. For example, motif analysis and motif location identification have been split up into distinct functions <code>find_motifs_genome</code> and <code>find_motifs_instances</code>, respectively.</p>
</div>
<div id="operation" class="section level2">
<h2 class="hasAnchor">
<a href="#operation" class="anchor"></a>Operation</h2>
<p><code>marge</code> is distributed on Bitbucket as an <code>R</code> package, and is compatible with Mac OS X, Windows, and Linux. The latest version of <a href="http://homer.ucsd.edu/homer/">HOMER</a> (v4.9 as of this writing) should be installed prior to <code>marge</code>, and <a href="http://homer.ucsd.edu/homer/introduction/configure.html">configured</a> with the desired genomes and feature sets. Subsequently, <code>marge</code> can be installed from the <a href="https://github.com/robertamezquita/marge/">repository</a>. <code>R</code> Package dependencies and system requirements are documented in the repository and automatically installed in conjunction with the <code>marge</code> package. <a href="https://marge.netlify.com/">Online documentation</a> is also available at <code>marge.netlify.com</code>.</p>
</div>
</div>
<div id="use-cases" class="section level1">
<h1 class="hasAnchor">
<a href="#use-cases" class="anchor"></a>Use Cases</h1>
<p>To demonstrate the functionality and utility of <code>marge</code>, we present a tutorial of core functionality and present potential use cases for high-throughput motif enrichment studies and subsequent downstream analyses.</p>
<div id="conventions" class="section level2">
<h2 class="hasAnchor">
<a href="#conventions" class="anchor"></a>Conventions</h2>
<p>Code objects shall be referred to in-line with the <code>code</code> font, whereas <code>R</code> functions will be referred to with a pair of parentheses following the function name, as in <code><a href="../reference/find_motifs_genome.html">find_motifs_genome()</a></code>. <code>R</code> code chunks will be presented as a block. Comments describing relevant portions of code are prepended by hashes (<code>#</code>), and all else is runnable <code>R</code> code. Results of code chunks are printed in separate block below, with each line of output prepended by a double hash (<code>##</code>). An example of code chunk and output formatting follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">salutation &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'hello world'</span>)</a>
<a class="sourceLine" id="cb1-2" title="2">salutation <span class="co"># view output</span></a></code></pre></div>
<pre><code>## [1] "hello world"</code></pre>
<p>Lastly, documentation for functions can be accessed invoking through help utility in <code>R</code>. <a href="https://marge.netlify.com">Online documentation</a> with the same content is also available under the ‘Reference’ tab.</p>
</div>
<div id="preamble" class="section level2">
<h2 class="hasAnchor">
<a href="#preamble" class="anchor"></a>Preamble</h2>
<p>The latest version of <a href="http://homer.ucsd.edu/homer/introduction/install.html">HOMER</a> should be installed prior to installing <code>marge</code> as documented online in the <a href="http://homer.ucsd.edu/homer/introduction/install.html">introduction/install</a> section. Following installation, HOMER packages should additionally be installed, which contains the necessary sequence data to perform motif enrichment, such as those that follow:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co">## In a terminal/command-line</span></a>
<a class="sourceLine" id="cb3-2" title="2">perl <span class="op">/</span>path<span class="op">-</span>to<span class="op">-</span>homer<span class="op">/</span>configureHomer.pl <span class="op">-</span>install hg38 <span class="co"># human sequence data</span></a>
<a class="sourceLine" id="cb3-3" title="3">perl <span class="op">/</span>path<span class="op">-</span>to<span class="op">-</span>homer<span class="op">/</span>configureHomer.pl <span class="op">-</span>install mm10 <span class="co"># mouse sequence data</span></a></code></pre></div>
<p>Additionally, users should take care to ensure that HOMER is added to the executable path, such that entering <code>findMotifsGenome.pl</code> on the command-line works as expected. For further details, we again refer to the <a href="http://homer.ucsd.edu/homer/introduction/install.html">introduction/install</a> section of the HOMER documentation.</p>
</div>
<div id="basic-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic Usage</h2>
<p>What follows is a basic tutorial of a typical workflow in performing motif enrichment analysis from a given, single set of regions. This section will introduce the various verbs available in <code>marge</code> for this task.</p>
<div id="check-homer-and-installed-packages" class="section level3">
<h3 class="hasAnchor">
<a href="#check-homer-and-installed-packages" class="anchor"></a>Check HOMER and Installed Packages</h3>
<p>To check that HOMER was installed properly and is detected by <code>marge</code> invoke the <code><a href="../reference/check_homer.html">check_homer()</a></code> function. Sequence annotation data for the genome of interest should also have been installed prior to running HOMER or <code>marge</code>. To check what is currently available via HOMER, simply run <code><a href="../reference/list_homer_packages.html">list_homer_packages()</a></code>. For this analysis, either ‘hg38’ or ‘mm10’ packages can be installed, which include the human and mouse genomes and annotations, respectively.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co">## Install `marge`:</span></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="co">## devtools::install_github('robertamezquita/marge', ref = 'master')</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(marge)</a>
<a class="sourceLine" id="cb4-4" title="4"></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="kw"><a href="../reference/check_homer.html">check_homer</a></span>()</a></code></pre></div>
<pre><code>## HOMER installation found.</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw"><a href="../reference/list_homer_packages.html">list_homer_packages</a></span>()</a></code></pre></div>
<pre><code>## # A tibble: 3 x 4
##   status package version description                                       
##   &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;                                             
## 1 +      homer   v4.10.3 Code/Executables, ontologies, motifs for HOMER    
## 2 +      mouse-o v6.0    Mus musculus (mouse) accession and ontology infor…
## 3 +      mm10    v6.0    mouse genome and annotation for UCSC mm10</code></pre>
</div>
<div id="input-data" class="section level3">
<h3 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input data</h3>
<p>To begin, regions of interest should first be identified. For example, peak calls derived from ATAC-seq or ChIP-seq data are of particular interest for motif enrichment analyses. At minimum, such regions should contain three columns: <code>chrom</code> (chromosome), <code>start</code>, and <code>end</code>, describing the location of the region of interest. Additional columns are also allowed - suggested columns include a region identifier, gene annotations, distance to the gene, and properties such as expression/chromatin state.</p>
<p>A minimal set of regions is included with the <code>marge</code> package for testing, and can be loaded as follows:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">## Use included test regions from `marge` package</span></a>
<a class="sourceLine" id="cb8-2" title="2">test_file &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'extdata/test_regions.bed'</span>, <span class="dt">package =</span> <span class="st">'marge'</span>)</a>
<a class="sourceLine" id="cb8-3" title="3">dat &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(test_file)</a>
<a class="sourceLine" id="cb8-4" title="4"></a>
<a class="sourceLine" id="cb8-5" title="5">dat</a></code></pre></div>
<pre><code>## # A tibble: 4 x 6
##   chrom     start       end id       value gene 
##   &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;
## 1 chr6  102207441 102207830 region_1     1 A    
## 2 chr7   44688352  44688664 region_2     2 B    
## 3 chr7   92093830  92094134 region_3     3 C    
## 4 chr12 103669461 103669829 region_4     4 D</code></pre>
</div>
<div id="find-motifs-across-the-genome" class="section level3">
<h3 class="hasAnchor">
<a href="#find-motifs-across-the-genome" class="anchor"></a>Find Motifs Across the Genome</h3>
<p>From the regions of interest, enriched <em>de novo</em> and/or known motifs can be performed via the <code><a href="../reference/find_motifs_genome.html">find_motifs_genome()</a></code> function. This function connects to the HOMER utility <code>findMotifsGenome.pl</code>, and uses much of the same syntax, and so should be familiar to users of HOMER. The following code shows all available options, with documentation available online and via <code>R</code> for further details. Suffice it to say, a <em>de novo</em> and known motifs enrichment analysis is run on the test regions loaded prior, and writes results to the designated path directory (here, <code>results_dir</code>). Thus, <code><a href="../reference/find_motifs_genome.html">find_motifs_genome()</a></code> is run purely for its side-effect of creating a HOMER results directory.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">## Create a temporary directory to write results</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="co">## This directory is erased once R session is closed</span></a>
<a class="sourceLine" id="cb10-3" title="3">results_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>(<span class="dt">pattern =</span> <span class="st">'test-dir_'</span>)</a></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co">## Run a motif enrichment analysis</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="kw"><a href="../reference/find_motifs_genome.html">find_motifs_genome</a></span>(</a>
<a class="sourceLine" id="cb11-3" title="3">    dat,</a>
<a class="sourceLine" id="cb11-4" title="4">    <span class="dt">path =</span> results_dir,</a>
<a class="sourceLine" id="cb11-5" title="5">    <span class="dt">genome =</span> <span class="st">'mm10'</span>, </a>
<a class="sourceLine" id="cb11-6" title="6">    <span class="dt">motif_length =</span> <span class="dv">8</span>,</a>
<a class="sourceLine" id="cb11-7" title="7">    <span class="dt">scan_size =</span> <span class="dv">50</span>,</a>
<a class="sourceLine" id="cb11-8" title="8">    <span class="dt">optimize_count =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb11-9" title="9">    <span class="dt">background =</span> <span class="st">'automatic'</span>,</a>
<a class="sourceLine" id="cb11-10" title="10">    <span class="dt">local_background =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb11-11" title="11">    <span class="dt">only_known =</span> <span class="ot">FALSE</span>, <span class="dt">only_denovo =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb11-12" title="12">    <span class="dt">fdr_num =</span> <span class="dv">5</span>,</a>
<a class="sourceLine" id="cb11-13" title="13">    <span class="dt">cores =</span> <span class="dv">1</span>, <span class="dt">cache =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb11-14" title="14">    <span class="dt">overwrite =</span> <span class="ot">TRUE</span>, <span class="dt">keep_minimal =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb11-15" title="15">)</a></code></pre></div>
<p>Note in particular the argument <code>keep_minimal</code>, which if set to <code>TRUE</code>, removes all extraneous HTML and logo images, retaining only the relevant enrichment results necessary for downstream analyses. If <code>FALSE</code> (the default), the output is exactly the same as if HOMER had been run from the command line with the given parameters.</p>
</div>
<div id="load-denovo-and-known-motif-enrichment-results" class="section level3">
<h3 class="hasAnchor">
<a href="#load-denovo-and-known-motif-enrichment-results" class="anchor"></a>Load Denovo and Known Motif Enrichment Results</h3>
<p>HOMER produces two distinct files for <em>de novo</em> and known motif enrichments within the designated results directory, contained in <code>homerMotifs.all.motif</code> and <code>knownResults.txt</code>, respectively. These results can be read in using the respective <code>read_*_results()</code> functions, where <code>*</code> is either <code>denovo</code> or <code>known</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">known  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_known_results.html">read_known_results</a></span>(<span class="dt">path =</span> results_dir, <span class="dt">homer_dir =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-2" title="2">denovo &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_denovo_results.html">read_denovo_results</a></span>(<span class="dt">path =</span> results_dir, <span class="dt">homer_dir =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-3" title="3"></a>
<a class="sourceLine" id="cb12-4" title="4">known</a>
<a class="sourceLine" id="cb12-5" title="5">denovo</a></code></pre></div>
<pre><code>## # A tibble: 365 x 14
##    motif_name motif_family experiment accession database consensus
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;    
##  1 Fosl2      bZIP         3T3L1-Fos… GSE56872  Homer    NATGASTC…
##  2 MafB       bZIP         BMM-Mafb-… GSE75722  Homer    WNTGCTGA…
##  3 Fra2       bZIP         Striatum-… GSE43429  Homer    GGATGACT…
##  4 Fra1       bZIP         BT549-Fra… GSE46166  Homer    NNATGAST…
##  5 JunB       bZIP         Dendritic… GSE36099  Homer    RATGASTC…
##  6 Atf3       bZIP         GBM-ATF3-… GSE33912  Homer    DATGASTC…
##  7 BATF       bZIP         Th17-BATF… GSE39756  Homer    DATGASTC…
##  8 AP-1       bZIP         ThioMac-P… GSE21512  Homer    VTGACTCA…
##  9 RUNX       Runt         HPC7-Runx… GSE22178  Homer    SAAACCAC…
## 10 NFY        CCAAT        Promoter   &lt;NA&gt;      Homer    RGCCAATS…
## # ... with 355 more rows, and 8 more variables: log_p_value &lt;dbl&gt;,
## #   fdr &lt;dbl&gt;, tgt_num &lt;dbl&gt;, tgt_pct &lt;dbl&gt;, bgd_num &lt;dbl&gt;, bgd_pct &lt;dbl&gt;,
## #   motif_pwm &lt;list&gt;, log_odds_detection &lt;dbl&gt;</code></pre>
<pre><code>## # A tibble: 2 x 17
##   consensus motif_name log_odds_detect… motif_pwm log_p_value_det… tgt_num
##   &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt; &lt;list&gt;               &lt;dbl&gt;   &lt;dbl&gt;
## 1 TCGCATTG  1-TCGCATTG             8.24 &lt;tibble …            -7.02       1
## 2 GATTACTC  2-GATTACTC             8.24 &lt;tibble …            -6.73       1
## # ... with 11 more variables: tgt_pct &lt;dbl&gt;, bgd_num &lt;dbl&gt;, bgd_pct &lt;dbl&gt;,
## #   log_p_value &lt;dbl&gt;, fdr &lt;dbl&gt;, tgt_pos &lt;dbl&gt;, tgt_std &lt;dbl&gt;,
## #   bgd_pos &lt;dbl&gt;, bgd_std &lt;dbl&gt;, strand_bias &lt;dbl&gt;, multiplicity &lt;dbl&gt;</code></pre>
<p>These functions specifically are responsible for the core routine which tidies the HOMER results, and can be pointed to any existing HOMER motif enrichment result files by setting <code>homer_dir</code> to <code>FALSE</code>. The output provides a tidied version of the default HOMER output, capturing the various characteristics of enrichments, such as the identity of the motifs, target and background number/percents of the motifs across the provided sequences, statistics of enrichment, and finally the associated motif position weight matrices (PWM) in the list-column <code>motif_pwm</code>. The columns are described further in the associated help files.</p>
</div>
<div id="accessing-motif-pwms" class="section level3">
<h3 class="hasAnchor">
<a href="#accessing-motif-pwms" class="anchor"></a>Accessing Motif PWMs</h3>
<p>One core task of interest is working with the identified <em>de novo</em> motif PWMs that were enriched across the regions of interest. To access a specific motif PWM, the following shows several ways of accessing them from the <code>motif_pwm</code> list-column in the <code>known</code> and <code>denovo</code> objects previously created.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="co">## Known Motifs</span></a>
<a class="sourceLine" id="cb15-2" title="2">known<span class="op">$</span>motif_pwm[<span class="dv">1</span>]             <span class="co"># positional access</span></a>
<a class="sourceLine" id="cb15-3" title="3">known<span class="op">$</span>motif_pwm[<span class="st">'OCT:OCT'</span>]     <span class="co"># named access</span></a>
<a class="sourceLine" id="cb15-4" title="4"></a>
<a class="sourceLine" id="cb15-5" title="5"><span class="co">## Denovo Motifs</span></a>
<a class="sourceLine" id="cb15-6" title="6">denovo<span class="op">$</span>motif_pwm[<span class="dv">1</span>]            <span class="co"># positional access</span></a></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">denovo<span class="op">$</span>motif_pwm[<span class="st">'1-TCGCATTG'</span>] <span class="co"># named access (denovo)</span></a></code></pre></div>
<pre><code>## $`1-TCGCATTG`
## # A tibble: 8 x 4
##       A     C     G     T
##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1   0.1   0.1   0.1   0.7
## 2   0.1   0.7   0.1   0.1
## 3   0.1   0.1   0.7   0.1
## 4   0.1   0.7   0.1   0.1
## 5   0.7   0.1   0.1   0.1
## 6   0.1   0.1   0.1   0.7
## 7   0.1   0.1   0.1   0.7
## 8   0.1   0.1   0.7   0.1</code></pre>
<p>Note the column names identifying the nucleotide, the rows for the ordered nucleotide position, and the frequencies at each cell, where each row sums to 1.</p>
</div>
<div id="writing-motif-pwms---manual" class="section level3">
<h3 class="hasAnchor">
<a href="#writing-motif-pwms---manual" class="anchor"></a>Writing Motif PWMs - Manual</h3>
<p>To write a motif to a file, the helper function <code><a href="../reference/write_homer_motif.html">write_homer_motif()</a></code> can be used, and values can be manually specified for motifs of interest.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co">## Note double bracket to access list value;</span></a>
<a class="sourceLine" id="cb18-2" title="2"><span class="co">## single bracket produces error</span></a>
<a class="sourceLine" id="cb18-3" title="3"><span class="kw"><a href="../reference/write_homer_motif.html">write_homer_motif</a></span>(</a>
<a class="sourceLine" id="cb18-4" title="4">    <span class="dt">motif_pwm =</span> denovo<span class="op">$</span>motif_pwm[[<span class="st">'1-TCGCATTG'</span>]], </a>
<a class="sourceLine" id="cb18-5" title="5">    <span class="dt">motif_name =</span> <span class="st">'my_first_motif'</span>,</a>
<a class="sourceLine" id="cb18-6" title="6">    <span class="dt">log_odds_detection =</span> <span class="fl">4.35</span>,</a>
<a class="sourceLine" id="cb18-7" title="7">    <span class="dt">consensus =</span> <span class="st">'CACATCCT'</span>,</a>
<a class="sourceLine" id="cb18-8" title="8">    <span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(results_dir, <span class="st">'/my_first_motif.motif'</span>)</a>
<a class="sourceLine" id="cb18-9" title="9">)</a></code></pre></div>
</div>
<div id="writing-motif-pwms---automatic" class="section level3">
<h3 class="hasAnchor">
<a href="#writing-motif-pwms---automatic" class="anchor"></a>Writing Motif PWMs - Automatic</h3>
<p>Oftentimes, one wants to write all associated motif PWMs from a given <em>de novo</em> enrichment search. Given that <code>write_homer_motif</code> produces a single motif at a time, we need a way to produce a file for each row of the <code>denovo</code> object.</p>
<p>While there are various ways to iterate over the rows of the <code>denovo</code> object, such as with <code>for-loops</code> or the <code>apply</code> family of functions, below is shown a method using the <a href="http://purrr.tidyverse.org/"><code>purrr</code></a> package from the <code>tidyverse</code> family, using the <code><a href="https://purrr.tidyverse.org/reference/map.html">map()</a></code> and <code><a href="https://purrr.tidyverse.org/reference/map.html">walk()</a></code> functions. Briefly, these functions apply a function to each element of a vector, thus allowing for iteration over the rows of an object such as those we created, <code>known</code> and <code>denovo</code>. <code>map</code> returns the transformed object, whereas <code>walk</code> calls the function for its side-effect, such as in the case of writing a file as in <code><a href="../reference/write_homer_motif.html">write_homer_motif()</a></code>. Given that we need several varying inputs (all the arguments of <code><a href="../reference/write_homer_motif.html">write_homer_motif()</a></code>, we use the variant <code><a href="https://purrr.tidyverse.org/reference/map2.html">pwalk()</a></code> to allow for our many varying inputs. The resulting code below writes all of our denovo results (three motifs in total) to unique files in our temporary <code>R</code> results directory.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="co">## Write to multiple files - vary the `file` argument</span></a>
<a class="sourceLine" id="cb19-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(purrr)</a>
<a class="sourceLine" id="cb19-3" title="3"><span class="kw"><a href="https://purrr.tidyverse.org/reference/map2.html">pwalk</a></span>(</a>
<a class="sourceLine" id="cb19-4" title="4">    <span class="dt">.l =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb19-5" title="5">        <span class="dt">motif_pwm  =</span> denovo<span class="op">$</span>motif_pwm,</a>
<a class="sourceLine" id="cb19-6" title="6">        <span class="dt">motif_name =</span> denovo<span class="op">$</span>motif_name,</a>
<a class="sourceLine" id="cb19-7" title="7">        <span class="dt">log_odds_detection =</span> denovo<span class="op">$</span>log_odds_detection,</a>
<a class="sourceLine" id="cb19-8" title="8">        <span class="dt">consensus =</span> denovo<span class="op">$</span>consensus,</a>
<a class="sourceLine" id="cb19-9" title="9">        <span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(results_dir, <span class="st">'/'</span>, denovo<span class="op">$</span>motif_name, <span class="st">'.motif'</span>)</a>
<a class="sourceLine" id="cb19-10" title="10">    ),</a>
<a class="sourceLine" id="cb19-11" title="11">    <span class="dt">.f =</span> write_homer_motif</a>
<a class="sourceLine" id="cb19-12" title="12">)</a>
<a class="sourceLine" id="cb19-13" title="13"></a>
<a class="sourceLine" id="cb19-14" title="14"><span class="co">## Write to a single file with all motifs</span></a>
<a class="sourceLine" id="cb19-15" title="15"><span class="co">## Keep the `file` argument constant</span></a>
<a class="sourceLine" id="cb19-16" title="16">my_motifs_file &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(results_dir, <span class="st">'/my-motifs.motif'</span>)</a>
<a class="sourceLine" id="cb19-17" title="17"></a>
<a class="sourceLine" id="cb19-18" title="18"><span class="co">## Append by default is TRUE, e.g. can write more than one</span></a>
<a class="sourceLine" id="cb19-19" title="19"><span class="co">## motif at once, but this means don't run this code more</span></a>
<a class="sourceLine" id="cb19-20" title="20"><span class="co">## than once!</span></a>
<a class="sourceLine" id="cb19-21" title="21"><span class="kw"><a href="https://purrr.tidyverse.org/reference/map2.html">pwalk</a></span>(</a>
<a class="sourceLine" id="cb19-22" title="22">    <span class="dt">.l =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb19-23" title="23">        <span class="dt">motif_pwm  =</span> denovo<span class="op">$</span>motif_pwm,</a>
<a class="sourceLine" id="cb19-24" title="24">        <span class="dt">motif_name =</span> denovo<span class="op">$</span>motif_name,</a>
<a class="sourceLine" id="cb19-25" title="25">        <span class="dt">log_odds_detection =</span> denovo<span class="op">$</span>log_odds_detection,</a>
<a class="sourceLine" id="cb19-26" title="26">        <span class="dt">consensus =</span> denovo<span class="op">$</span>consensus</a>
<a class="sourceLine" id="cb19-27" title="27">     ),</a>
<a class="sourceLine" id="cb19-28" title="28">    <span class="dt">.f =</span> write_homer_motif,</a>
<a class="sourceLine" id="cb19-29" title="29">    <span class="dt">file =</span> my_motifs_file,</a>
<a class="sourceLine" id="cb19-30" title="30">    <span class="dt">append =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb19-31" title="31">)</a></code></pre></div>
</div>
<div id="search-for-instances-of-motifs-across-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#search-for-instances-of-motifs-across-regions" class="anchor"></a>Search for Instances of Motifs Across Regions</h3>
<p>Now that we have produced our motif PWM as a file, it can be used to map the motif back to where it occurs in our set of regions using the <code><a href="../reference/find_motifs_instances.html">find_motifs_instances()</a></code> function. While it is also a wrapper for <code>findMotifsGenome.pl</code>, it partitions out a unique piece of functionality that presents an entirely different type of results. Note that this function <em>requires</em> the use of a HOMER motif file that is written out (either by hand or by the <code><a href="../reference/write_homer_motif.html">write_homer_motif()</a></code> function) as it contains the necessary parametrization for motif finding.</p>
<p>Also note that we can search for either a single motif or multiple motifs concurrently, where the multiple motifs are written into a single file, as shown above. Here, we will search for all three of the <em>de novo</em> motifs identified initially in our regions.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">motif_instances_file &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(results_dir, <span class="st">'/motif-instances.txt'</span>)</a>
<a class="sourceLine" id="cb20-2" title="2"></a>
<a class="sourceLine" id="cb20-3" title="3"><span class="kw"><a href="../reference/find_motifs_instances.html">find_motifs_instances</a></span>(</a>
<a class="sourceLine" id="cb20-4" title="4">    dat,</a>
<a class="sourceLine" id="cb20-5" title="5">    <span class="dt">path =</span> motif_instances_file,</a>
<a class="sourceLine" id="cb20-6" title="6">    <span class="dt">genome =</span> <span class="st">'mm10'</span>,</a>
<a class="sourceLine" id="cb20-7" title="7">    <span class="dt">motif_file =</span> my_motifs_file,</a>
<a class="sourceLine" id="cb20-8" title="8">    <span class="dt">scan_size =</span> <span class="dv">50</span>,</a>
<a class="sourceLine" id="cb20-9" title="9">    <span class="dt">cores =</span> <span class="dv">1</span>, <span class="dt">cache =</span> <span class="dv">100</span></a>
<a class="sourceLine" id="cb20-10" title="10">)</a></code></pre></div>
<p>This function, similarly to <code><a href="../reference/find_motifs_genome.html">find_motifs_genome()</a></code> and <code><a href="../reference/write_homer_motif.html">write_homer_motif()</a></code>, produces output that is written to into a file outside of <code>R</code>, and similarly relies on a helper function to read the resulting output.</p>
</div>
<div id="read-instances-of-motifs-across-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#read-instances-of-motifs-across-regions" class="anchor"></a>Read Instances of Motifs Across Regions</h3>
<p>The function <code><a href="../reference/read_motifs_instances.html">read_motifs_instances()</a></code> works in tandem with <code><a href="../reference/find_motifs_instances.html">find_motifs_instances()</a></code> to read its output, returning the location and scores of the specified motifs from the previous step.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">motif_instances &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_motifs_instances.html">read_motifs_instances</a></span>(motif_instances_file)</a>
<a class="sourceLine" id="cb21-2" title="2"></a>
<a class="sourceLine" id="cb21-3" title="3">motif_instances</a></code></pre></div>
<pre><code>## # A tibble: 2 x 6
##   region_id offset sequence motif_name strand motif_score
##   &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;        &lt;dbl&gt;
## 1 region_2     -24 TCGCATTG 1-TCGCATTG +             8.24
## 2 region_1      -5 GATTACTC 2-GATTACTC +             8.24</code></pre>
</div>
</div>
<div id="advanced-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-usage" class="anchor"></a>Advanced Usage</h2>
<p>Using the verbs learned in the previous section, this part will focus on tying it all together to create a framework for testing for motif enrichment and subsequently analysing the results across multiple sets of regions in a tidy fashion. This section relies heavily on the <code>tidyverse</code> framework, and in particular will use the tools from <code>purrr</code>, <code>map()/walk()</code>, introduced in the previous section. Additionally, we will simulate a dataset using the <code>valr</code> package function <code><a href="http://valr.hesselberthlab.org/reference/bed_random.html">bed_random()</a></code>. While outside of the scope of this tutorial, <code>valr</code> provides additional functionality to work with genomic intervals using similar <code>tidyverse</code> conventions, and is recommended for manipulating regions of interest from high-throughput experiments.</p>
<p>Below are the libraries required for this next section.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="co">## Required libraries for advanced usage tutorial</span></a>
<a class="sourceLine" id="cb23-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(valr)</a>
<a class="sourceLine" id="cb23-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb23-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyr)</a>
<a class="sourceLine" id="cb23-5" title="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(purrr)</a>
<a class="sourceLine" id="cb23-6" title="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a></code></pre></div>
<div id="constructing-multiple-simulated-sets-of-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#constructing-multiple-simulated-sets-of-regions" class="anchor"></a>Constructing Multiple Simulated Sets of Regions</h3>
<p>By bringing the usage of HOMER into <code>R</code> via an API, it becomes possible to construct multiple queries, execute them all, and then read in the results from all the queries.</p>
<p>First, we will create a simulated set of intervals for which to test for motif enrichment, formatting the regions into a <code>tibble</code> object which allows for the creation of list-columns, which tidily encapsulate the genomic intervals.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1"><span class="co">## 1. Construct multiple region samples -------------------------------------</span></a>
<a class="sourceLine" id="cb24-2" title="2"><span class="co">## Create genome from first two chromosomes of mouse</span></a>
<a class="sourceLine" id="cb24-3" title="3">genome &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">chrom =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'chr1'</span>, <span class="st">'chr2'</span>),</a>
<a class="sourceLine" id="cb24-4" title="4">                     <span class="dt">size =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">195471971</span>, <span class="dv">182113224</span>))</a>
<a class="sourceLine" id="cb24-5" title="5"></a>
<a class="sourceLine" id="cb24-6" title="6"><span class="co">## Set seeds and ids for the simulated sets of regions</span></a>
<a class="sourceLine" id="cb24-7" title="7">seed &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">5</span></a>
<a class="sourceLine" id="cb24-8" title="8">id &lt;-<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb24-9" title="9"></a>
<a class="sourceLine" id="cb24-10" title="10"><span class="co">## Simulate different sets of regions (each with unique seeds) and munge</span></a>
<a class="sourceLine" id="cb24-11" title="11"><span class="co">## into the tibble format, keeping the regions as a list-column</span></a>
<a class="sourceLine" id="cb24-12" title="12">sim &lt;-<span class="st"> </span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(seed, <span class="op">~</span><span class="st"> </span>valr<span class="op">::</span><span class="kw"><a href="http://valr.hesselberthlab.org/reference/bed_random.html">bed_random</a></span>(genome, <span class="dt">length =</span> <span class="dv">50</span>, <span class="dt">n =</span> <span class="dv">50</span>, <span class="dt">seed =</span> .x))</a>
<a class="sourceLine" id="cb24-13" title="13">tbl_regions &lt;-<span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/reexports.html">tibble</a></span>(<span class="dt">id =</span> id, <span class="dt">regions =</span> sim)</a>
<a class="sourceLine" id="cb24-14" title="14"></a>
<a class="sourceLine" id="cb24-15" title="15"><span class="co">## Inspect overall tibble organization</span></a>
<a class="sourceLine" id="cb24-16" title="16">tbl_regions</a></code></pre></div>
<pre><code>## # A tibble: 5 x 2
##   id    regions          
##   &lt;chr&gt; &lt;list&gt;           
## 1 a     &lt;tibble [50 × 3]&gt;
## 2 b     &lt;tibble [50 × 3]&gt;
## 3 c     &lt;tibble [50 × 3]&gt;
## 4 d     &lt;tibble [50 × 3]&gt;
## 5 e     &lt;tibble [50 × 3]&gt;</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="co">## View regions sample</span></a>
<a class="sourceLine" id="cb26-2" title="2">tbl_regions<span class="op">$</span>regions[[<span class="dv">1</span>]] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="dt">n =</span> <span class="dv">5</span>)</a></code></pre></div>
<pre><code>## # A tibble: 50 x 3
##   chrom    start      end
##   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;
## 1 chr1   3322213  3322263
## 2 chr1  12952049 12952099
## 3 chr1  13419401 13419451
## 4 chr1  13496953 13497003
## 5 chr1  44639848 44639898
## # ... with 45 more rows</code></pre>
</div>
<div id="finding-motifs-across-multiple-sets-of-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#finding-motifs-across-multiple-sets-of-regions" class="anchor"></a>Finding Motifs Across Multiple Sets of Regions</h3>
<p>The next step involves specifying the paths to the results directory for each respective set of regions. This is necessary for <code><a href="../reference/find_motifs_genome.html">find_motifs_genome()</a></code> to output the motif enrichment results for each set of regions separately (otherwise the results will be overwritten multiple times).</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="co">## 2. Find Motifs in Genome -------------------------------------------------</span></a>
<a class="sourceLine" id="cb28-2" title="2"></a>
<a class="sourceLine" id="cb28-3" title="3"><span class="co">## Create new (temp) results directory</span></a>
<a class="sourceLine" id="cb28-4" title="4">results_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>(<span class="dt">pattern =</span> <span class="st">'ht-dir_'</span>)</a>
<a class="sourceLine" id="cb28-5" title="5"></a>
<a class="sourceLine" id="cb28-6" title="6"><span class="co">## Append results path to tibble of regions</span></a>
<a class="sourceLine" id="cb28-7" title="7">tbl_regions &lt;-<span class="st"> </span>tbl_regions <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb28-8" title="8"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(results_dir, <span class="st">'/homer_'</span>, id))</a>
<a class="sourceLine" id="cb28-9" title="9"></a>
<a class="sourceLine" id="cb28-10" title="10">tbl_regions[<span class="dv">1</span>, ] <span class="co"># inspect resulting first row</span></a></code></pre></div>
<pre><code>## # A tibble: 1 x 3
##   id    regions        path                                                
##   &lt;chr&gt; &lt;list&gt;         &lt;chr&gt;                                               
## 1 a     &lt;tibble [50 ×… /var/folders/8t/qchxz4r950zd5qnp21glqt8w0000gp/T//R…</code></pre>
<p>Now, using each set of regions and the unique paths for each set of regions, <code><a href="../reference/find_motifs_genome.html">find_motifs_genome()</a></code> can be run, varying across these two parameters, and setting as constant the various motif enrichment search parameters for all the different sets.</p>
<p>One other potential application not shown here is testing how these various parameters affect motif enrichment results if we were instead to keep the regions tested as constant instead. This is left as an exercise for the interested reader.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="co">## Iterate over all regions to perform motif finding</span></a>
<a class="sourceLine" id="cb30-2" title="2"><span class="co">## Writing each set of regions results to different directory</span></a>
<a class="sourceLine" id="cb30-3" title="3"><span class="co">## Only perform known motif results enrichments</span></a>
<a class="sourceLine" id="cb30-4" title="4"><span class="kw"><a href="https://purrr.tidyverse.org/reference/map2.html">pwalk</a></span>(</a>
<a class="sourceLine" id="cb30-5" title="5">    <span class="co">## Varying parameters (regions, output directory)</span></a>
<a class="sourceLine" id="cb30-6" title="6">    <span class="dt">.l =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">x =</span> tbl_regions<span class="op">$</span>regions, <span class="dt">path =</span> tbl_regions<span class="op">$</span>path),</a>
<a class="sourceLine" id="cb30-7" title="7">    <span class="co">## Function to find motifs across genome</span></a>
<a class="sourceLine" id="cb30-8" title="8">    <span class="dt">.f =</span> find_motifs_genome,</a>
<a class="sourceLine" id="cb30-9" title="9">    <span class="co">## Constant parameters (motif search settings)</span></a>
<a class="sourceLine" id="cb30-10" title="10">    <span class="dt">genome =</span> <span class="st">'mm10'</span>,</a>
<a class="sourceLine" id="cb30-11" title="11">    <span class="dt">scan_size =</span> <span class="dv">50</span>,</a>
<a class="sourceLine" id="cb30-12" title="12">    <span class="dt">optimize_count =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb30-13" title="13">    <span class="dt">only_known =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb30-14" title="14">    <span class="dt">cores =</span> <span class="dv">2</span>, <span class="dt">cache =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb30-15" title="15">    <span class="dt">keep_minimal =</span> <span class="ot">TRUE</span>, <span class="dt">overwrite =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb30-16" title="16">)</a></code></pre></div>
</div>
<div id="combining-results-from-multiple-motif-enrichment-searches" class="section level3">
<h3 class="hasAnchor">
<a href="#combining-results-from-multiple-motif-enrichment-searches" class="anchor"></a>Combining Results from Multiple Motif Enrichment Searches</h3>
<p>Following the motif enrichment search for each set of regions, the results can be read in, iterating over the variable <code>path</code> within the <code>tbl_regions</code> object to pull in the relevant results. The results are cleaned to remove the <code>path</code> and <code>regions</code> columns subsequently, and then expanded such that each individual motif enrichment result has its corresponding <code>id</code> as a separate variable.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1"><span class="co">## 3. Read in All Results ----------------------------------------------------</span></a>
<a class="sourceLine" id="cb31-2" title="2"></a>
<a class="sourceLine" id="cb31-3" title="3"><span class="co">## Read known motif enrichment results into a new list-column</span></a>
<a class="sourceLine" id="cb31-4" title="4"><span class="co">## Remove the path and regions columns for cleanliness</span></a>
<a class="sourceLine" id="cb31-5" title="5"><span class="co">## Unnest results to expand into plotting friendly format</span></a>
<a class="sourceLine" id="cb31-6" title="6">tbl_results &lt;-<span class="st"> </span>tbl_regions <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb31-7" title="7"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">known_results =</span> <span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(path, read_known_results)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb31-8" title="8"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span>path, <span class="op">-</span>regions) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb31-9" title="9"><span class="st">    </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/unnest.html">unnest</a></span>()</a></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1"><span class="co">## View sampling of results</span></a>
<a class="sourceLine" id="cb32-2" title="2">tbl_results <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/sample.html">sample_n</a></span>(<span class="dt">size =</span> <span class="dv">5</span>)</a></code></pre></div>
<pre><code>## # A tibble: 5 x 15
##   id    motif_name motif_family experiment accession database consensus
##   &lt;chr&gt; &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;    
## 1 c     JunD       bZIP         K562-JunD… &lt;NA&gt;      Homer    ATGACGTC…
## 2 d     ARE        NR           LNCAP-AR-… GSE27824  Homer    RGRACASN…
## 3 d     E-box      bHLH         Promoter   &lt;NA&gt;      Homer    SSGGTCAC…
## 4 a     ETS:RUNX   ETS,Runt     Jurkat-RU… GSE17954  Homer    RCAGGATG…
## 5 b     Srebp2     bHLH         HepG2-Sre… GSE31477  Homer    CGGTCACS…
## # ... with 8 more variables: log_p_value &lt;dbl&gt;, fdr &lt;dbl&gt;, tgt_num &lt;dbl&gt;,
## #   tgt_pct &lt;dbl&gt;, bgd_num &lt;dbl&gt;, bgd_pct &lt;dbl&gt;, motif_pwm &lt;list&gt;,
## #   log_odds_detection &lt;dbl&gt;</code></pre>
</div>
<div id="summarising-characteristics-of-motif-enrichment-results" class="section level3">
<h3 class="hasAnchor">
<a href="#summarising-characteristics-of-motif-enrichment-results" class="anchor"></a>Summarising Characteristics of Motif Enrichment Results</h3>
<p>In this tidy form, it becomes possible to summarise results on the basis of motif families, as opposed to simply individual motifs. Below, we use the <code>id</code> and <code>motif_family</code> columns to group the results, and summarise a given motif family by taking the most significant individual motif appearance as the summary statistic (e.g., the max -log10 p-value).</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1"><span class="co">## Summarise number of </span></a>
<a class="sourceLine" id="cb34-2" title="2">tbl_summary &lt;-<span class="st"> </span>tbl_results <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb34-3" title="3"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(id, motif_family) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb34-4" title="4"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">top_log_p_value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Extremes">max</a></span>(log_p_value)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb34-5" title="5"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span>()</a>
<a class="sourceLine" id="cb34-6" title="6"></a>
<a class="sourceLine" id="cb34-7" title="7"><span class="co">## Print example results</span></a>
<a class="sourceLine" id="cb34-8" title="8">tbl_summary <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/sample.html">sample_n</a></span>(<span class="dv">5</span>)</a></code></pre></div>
<pre><code>## # A tibble: 5 x 3
##   id    motif_family  top_log_p_value
##   &lt;chr&gt; &lt;chr&gt;                   &lt;dbl&gt;
## 1 d     NR,IR1                      0
## 2 d     CTF                         0
## 3 a     ETS:IRF                     0
## 4 a     Homeobox,bHLH               0
## 5 c     E2F                         0</code></pre>
<p>Note that few motifs appear significant in this simulated dataset, as regions were chosen at random, without a true biological correlate.</p>
</div>
<div id="plotting-results" class="section level3">
<h3 class="hasAnchor">
<a href="#plotting-results" class="anchor"></a>Plotting Results</h3>
<p>Finally, we can inspect the results of our analyses using <code>ggplot2</code> package conventions, given the tidiness of the results. Here we show a subset of the overall results for succinctness.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1">families &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'AP2'</span>, <span class="st">'bHLH'</span>, <span class="st">'bZIP'</span>, <span class="st">'CCAAT'</span>, <span class="st">'CP2'</span>)</a>
<a class="sourceLine" id="cb36-2" title="2"></a>
<a class="sourceLine" id="cb36-3" title="3">tbl_summary <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb36-4" title="4"><span class="st">    </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(motif_family <span class="op">%in%</span><span class="st"> </span>families) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb36-5" title="5"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> motif_family, <span class="dt">y =</span> id, <span class="dt">fill =</span> top_log_p_value)) <span class="op">+</span></a>
<a class="sourceLine" id="cb36-6" title="6"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>()</a></code></pre></div>
<p><img src="marge-workflow_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
</div>
</div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p><code>marge</code> presents an <code>R</code> centric form of performing and analyzing motif enrichment results using the popular HOMER suite of tools, providing an easy-to-use interface and results in line with modern <code>R</code> idioms. We envision that <code>marge</code> will serve as a valuable tool to assist researchers in performing motif enrichment analyses quickly, easily, and reproducibly.</p>
</div>
<div id="data-and-software-availability" class="section level1">
<h1 class="hasAnchor">
<a href="#data-and-software-availability" class="anchor"></a>Data and Software Availability</h1>
<p>Online documentation for HOMER and installation details can be found at: <a href="http://homer.ucsd.edu/homer/" class="uri">http://homer.ucsd.edu/homer/</a></p>
<p><code>marge</code> can be installed via <code>devtools</code> using: <code><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">devtools::install_github('robertamezquita/marge', ref = 'master')</a></code></p>
<p>The latest <code>marge</code> source code is available at: <code>https://github.com/robertamezquita/marge/</code></p>
<p>Stable versions of <code>marge</code> are located in the Downloads section of the source code repository, under the tab ‘Tags’, at: <code>https://github.com/robertamezquita/marge/downloads/</code></p>
<!-- Online documentation for `marge` can be found at: `https://marge.netlify.com/` -->
</div>
<div id="competing-interests" class="section level1">
<h1 class="hasAnchor">
<a href="#competing-interests" class="anchor"></a>Competing Interests</h1>
<p>No competing interests were disclosed.</p>
</div>
<div id="acknowledgements" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h1>
<p>R.A.A. wrote the vignette, designed and implemented the software package. Jason Vander Heiden edited the manuscript and reviewed source code for clarity. Greg Finak contributed to code sussing out HOMER installation whereabouts.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#methods">Methods</a><ul class="nav nav-pills nav-stacked">
<li><a href="#implementation">Implementation</a></li>
      <li><a href="#operation">Operation</a></li>
      </ul>
</li>
      <li>
<a href="#use-cases">Use Cases</a><ul class="nav nav-pills nav-stacked">
<li><a href="#conventions">Conventions</a></li>
      <li><a href="#preamble">Preamble</a></li>
      <li><a href="#basic-usage">Basic Usage</a></li>
      <li><a href="#advanced-usage">Advanced Usage</a></li>
      </ul>
</li>
      <li><a href="#summary">Summary</a></li>
      <li><a href="#data-and-software-availability">Data and Software Availability</a></li>
      <li><a href="#competing-interests">Competing Interests</a></li>
      <li><a href="#acknowledgements">Acknowledgements</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Robert A. Amezquita.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
