<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find Motifs over Regions — find_motifs_genome • marge</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Find Motifs over Regions — find_motifs_genome" />
<meta property="og:description" content="Calls findMotifsGenome.pl to run motif analysis over a given
set of regions." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">marge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4.9999</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/marge-workflow.html">`marge`: An API for Analysis of Motifs Using HOMER in R</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find Motifs over Regions</h1>
    
    <div class="hidden name"><code>find_motifs_genome.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Calls <code>findMotifsGenome.pl</code> to run motif analysis over a given
set of regions.</p>
    </div>

    <pre class="usage"><span class='fu'>find_motifs_genome</span>(<span class='no'>x</span>, <span class='no'>path</span>, <span class='no'>genome</span>, <span class='kw'>motif_length</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>8</span>, <span class='fl'>10</span>, <span class='fl'>12</span>),
  <span class='kw'>scan_size</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>optimize_count</span> <span class='kw'>=</span> <span class='fl'>8</span>, <span class='kw'>background</span> <span class='kw'>=</span> <span class='st'>"automatic"</span>,
  <span class='kw'>local_background</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>only_known</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>only_denovo</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>fdr_num</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>cores</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span>(),
  <span class='kw'>cache</span> <span class='kw'>=</span> <span class='fu'>.calc_free_mem</span>()/<span class='fl'>4</span>, <span class='kw'>overwrite</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>keep_minimal</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>data.frame</code> with the first three columns being
chromosome, start, and end coordinates, with a fourth column corresponding
to a region identifier; extra columns may be kept; x may alternately be a
path to an existing bed file of this format</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>where to write HOMER results</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>ID of installed genome; check installed genomes using
<code><a href='list_homer_packages.html'>list_homer_packages()</a></code>; examples include "hg38" and "mm10";
add an 'r' at the end to mask repeats, e.g. "mm10r"</p></td>
    </tr>
    <tr>
      <th>motif_length</th>
      <td><p>vector of motif lengths to consider [default is
<code><a href='https://rdrr.io/r/base/c.html'>c(8, 10, 12)</a></code>]</p></td>
    </tr>
    <tr>
      <th>scan_size</th>
      <td><p>size of sequence to scan; this can be a numeric to
specify the number of bases to scan centered on the region, or
alternately can be set to "given" to scan the entire region;
if using "given", will use the "-chopify" option to cut large
background sequences to average of target sequence size
[default: <code>100</code>]</p></td>
    </tr>
    <tr>
      <th>optimize_count</th>
      <td><p>number of motifs to optimize [default: 8]</p></td>
    </tr>
    <tr>
      <th>background</th>
      <td><p><code>data.frame</code> containing coordinates of desired
regions to use as the background; alternately may be a path to an
existing bedfile; the default, "automatic", creates a background based
on the GC content and (scan) size of the target sequences</p></td>
    </tr>
    <tr>
      <th>local_background</th>
      <td><p>a numeric scalar specifying number of equal size
regions around peaks to use as the local background
[by default this is not used, e.g. default: <code>FALSE</code>]</p></td>
    </tr>
    <tr>
      <th>only_known</th>
      <td><p>turns off searching for denovo motifs</p></td>
    </tr>
    <tr>
      <th>only_denovo</th>
      <td><p>turns off searching for known motif enrichment</p></td>
    </tr>
    <tr>
      <th>fdr_num</th>
      <td><p>number of randomizations to perform to calculate FDR [default: 0]</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of cores to use [default: all cores available]</p></td>
    </tr>
    <tr>
      <th>cache</th>
      <td><p>number in MB to use as cache to store sequences in memory
[default: calculates free memory and divides by 4]</p></td>
    </tr>
    <tr>
      <th>overwrite</th>
      <td><p>overwrite an existing HOMER results directory
[default: <code>FALSE</code>]</p></td>
    </tr>
    <tr>
      <th>keep_minimal</th>
      <td><p>remove all extra clutter from results, keep only
the essentials (<code>knownResults.txt</code> and <code>homerMotifs.all.motifs</code>
[default: <code>FALSE</code>]</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Nothing; called for its side-effect of producing HOMER results</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>find_motifs_genome</code> runs the core HOMER motif enrichment function
from the R system, and in the process generates (as a side-effect)
a HOMER results directory.</p>
<p>This results directory is inspectable via a file browser, and contains
a summary of the results as HTML files as well as text files.</p>
<p>For our purposes, within the directory two key files exist:</p>
<ul>
<li><p><code>knownResults.txt</code> known motifs that are enriched</p></li>
<li><p><code>homerResults.all.motif</code> denovo motifs that are enriched</p></li>
</ul>

<p>These two text files are the core results (all else can be discarded
by setting <code>keep_minimal</code> to <code>TRUE</code>, and are parsed downstream
by <code><a href='read_known_results.html'>read_known_results</a></code> and <code><a href='read_denovo_results.html'>read_denovo_results</a></code>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='read_known_results.html'>read_known_results</a></code>, <code><a href='read_denovo_results.html'>read_denovo_results</a></code></p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robert A. Amezquita.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


